tests/python/pants_test/cache/test_pinger.py

```diff
@@ -4,68 +4,80 @@
 
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+import time
+
+import requests
+import responses
 from future.moves.urllib.parse import urlparse
 from requests import RequestException
 
 from pants.cache.pinger import BestUrlSelector, InvalidRESTfulCacheProtoError, Pinger
-from pants_test.cache.delay_server import setup_delayed_server
 from pants_test.test_base import TestBase
 
 
 class TestPinger(TestBase):
 
-  resolution = 1
-  fast_delay_seconds = 0
-  fast_timeout_seconds = fast_delay_seconds + resolution
-  slow_delay_seconds = fast_timeout_seconds + resolution
-  slow_timeout_seconds = slow_delay_seconds + resolution
-  unreachable_delay_seconds = slow_timeout_seconds + 10 * resolution
-  message = "This test may fail occasionally if the CPU is very busy."
+  fast_url = 'http://fast_url'
+  slow_url = 'http://slow_url'
+  unreachable_url = 'http://unreachable_url'
 
-  def setUp(self):
-    unreachable = setup_delayed_server(self.unreachable_delay_seconds)
-    slow = setup_delayed_server(self.slow_delay_seconds)
-    fast = setup_delayed_server(self.fast_delay_seconds)
-    self.servers = [unreachable, slow, fast]
-    self.fast_netloc = 'http://localhost:{}'.format(fast.socket.getsockname()[1])
-    self.slow_netloc = 'http://localhost:{}'.format(slow.socket.getsockname()[1])
-    self.unreachable_netloc = 'http://localhost:{}'.format(unreachable.socket.getsockname()[1])
-    self.https_external_netlock = 'https://github.com'
+  latency_by_url = {
+    fast_url: 0.1,
+    slow_url: 0.3,
+    unreachable_url: Pinger.UNREACHABLE
+  }
 
-  def test_pinger_times_correct(self):
-    test = Pinger(timeout=self.slow_timeout_seconds, tries=2)
-    netlocs = [self.fast_netloc, self.slow_netloc, self.unreachable_netloc]
-    ping_results = dict(test.pings(netlocs))
-    self.assertNotEqual(ping_results[self.slow_netloc], Pinger.UNREACHABLE)
-    self.assertLess(ping_results[self.fast_netloc], ping_results[self.slow_netloc])
-    self.assertEqual(ping_results[self.unreachable_netloc], Pinger.UNREACHABLE, msg=self.message)
+  @classmethod
+  def expect_response(cls, url, timeout):
+    latency = cls.latency_by_url[url]
+    body = requests.exceptions.ConnectTimeout() if latency >= timeout else ''
+
+    def callback(_):
+      if latency < timeout:
+        time.sleep(latency)
```

Comment:
I believe this will result in the same errors as the original. The flake in these tests was always due to `time.sleep` behaving unpredictably on CI machines under heavy load.

Reply:
Thanks - that makes sense. This is only needed for 1 of the test methods - let me see what I can do to eliminate this outright.

Reply:
Fixed - PTAL